File: triptracking-library/android/app/src/main/java/com/carmd/triptracking/api/TripTrackerAPIService.java
package com.carmd.triptracking.api;

import android.content.Context;
import android.location.Location;
import android.net.ConnectivityManager;
import android.net.Network;
import android.net.NetworkCapabilities;
import android.net.NetworkRequest;
import android.os.Build;
import android.util.Log;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.*;
import java.net.HttpURLConnection;
import java.net.URL;
import java.text.SimpleDateFormat;
import java.util.*;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public final class TripTrackerAPIService {
    private static final String TAG = "TripTrackerAPI";
    private static final int MAX_QUEUE_SIZE = 500;
    private static TripTrackerAPIService instance;

    // Config
    private String pingURL = "";
    private String endURL = "";
    private String userId = "";
    private String vehicleId = "";           // Optional
    private String osInfo = "Android " + Build.VERSION.RELEASE;
    private String routeId = "";
    private String authorizationKey = "";
    private String apiAuthKey = "";          // Legacy
    private String apiAuthToken = "";        // New header: api-auth-token

    // Whether to include vehicle_id in outgoing payloads
    // True during active trip, false otherwise
    private boolean includeVehicleId = false;

    private final ExecutorService executor = Executors.newSingleThreadExecutor();

    // ═══════════════════════════════════════════════════════════════
    // Retry Queue — persists failed requests to disk
    // ═══════════════════════════════════════════════════════════════

    private final CopyOnWriteArrayList<String> pendingQueue = new CopyOnWriteArrayList<>();
    private volatile boolean isFlushing = false;
    private Context appContext;
    private ConnectivityManager.NetworkCallback networkCallback;

    private TripTrackerAPIService() {}

    public static synchronized TripTrackerAPIService getInstance() {
        if (instance == null) instance = new TripTrackerAPIService();
        return instance;
    }

    /** Call once with app context to enable queue persistence + network monitoring */
    public void setContext(Context ctx) {
        this.appContext = ctx.getApplicationContext();
        loadPendingQueue();
        startNetworkMonitor();
    }

    // ═══════════════════════════════════════════════════════════════
    // Queue Persistence
    // ═══════════════════════════════════════════════════════════════

    private File getQueueFile() {
        if (appContext == null) return null;
        return new File(appContext.getCacheDir(), "triptracker_pending_api.json");
    }

    private void loadPendingQueue() {
        File file = getQueueFile();
        if (file == null || !file.exists()) return;
        try {
            BufferedReader reader = new BufferedReader(new FileReader(file));
            StringBuilder sb = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) sb.append(line);
            reader.close();

            JSONArray arr = new JSONArray(sb.toString());
            pendingQueue.clear();
            for (int i = 0; i < arr.length(); i++) {
                pendingQueue.add(arr.getString(i));
            }
            Log.i(TAG, "Queue loaded: " + pendingQueue.size() + " pending requests");
        } catch (Exception e) {
            Log.e(TAG, "Load queue error: " + e.getMessage());
        }
    }

    private void savePendingQueue() {
        File file = getQueueFile();
        if (file == null) return;
        try {
            JSONArray arr = new JSONArray();
            for (String item : pendingQueue) arr.put(item);
            FileWriter writer = new FileWriter(file);
            writer.write(arr.toString());
            writer.close();
        } catch (Exception e) {
            Log.e(TAG, "Save queue error: " + e.getMessage());
        }
    }

    private void enqueue(String url, JSONObject body) {
        try {
            JSONObject item = new JSONObject();
            item.put("url", url);
            item.put("body", body.toString());
            item.put("ts", System.currentTimeMillis());
            pendingQueue.add(item.toString());

            // Trim oldest if over limit
            while (pendingQueue.size() > MAX_QUEUE_SIZE) {
                pendingQueue.remove(0);
            }
            savePendingQueue();
            Log.i(TAG, "Queued (total: " + pendingQueue.size() + ") — will retry when online");
        } catch (Exception e) {
            Log.e(TAG, "Enqueue error: " + e.getMessage());
        }
    }

    /** Flush all pending requests. Called when network becomes available. */
    public void flushQueue() {
        if (isFlushing || pendingQueue.isEmpty()) return;
        isFlushing = true;

        executor.execute(() -> {
            int sent = 0;
            Log.i(TAG, "Flushing " + pendingQueue.size() + " pending requests…");

            Iterator<String> it = pendingQueue.iterator();
            while (it.hasNext()) {
                try {
                    JSONObject item = new JSONObject(it.next());
                    String url = item.getString("url");
                    JSONObject body = new JSONObject(item.getString("body"));

                    boolean ok = post(url, body);
                    if (ok) {
                        it.remove();
                        sent++;
                    } else {
                        break;  // Still offline — stop flushing
                    }
                } catch (Exception e) {
                    it.remove();  // Corrupt entry — remove
                }
            }
            savePendingQueue();
            isFlushing = false;
            Log.i(TAG, "Flush done: " + sent + " sent, " + pendingQueue.size() + " remaining");
        });
    }

    public int getPendingCount() { return pendingQueue.size(); }

    // ═══════════════════════════════════════════════════════════════
    // Network Monitor — auto-flush when connectivity returns
    // ═══════════════════════════════════════════════════════════════

    private void startNetworkMonitor() {
        if (appContext == null) return;
        try {
            ConnectivityManager cm = (ConnectivityManager) appContext.getSystemService(Context.CONNECTIVITY_SERVICE);
            if (cm == null) return;

            networkCallback = new ConnectivityManager.NetworkCallback() {
                @Override
                public void onAvailable(Network network) {
                    if (!pendingQueue.isEmpty()) {
                        Log.i(TAG, "Network restored — flushing pending queue");
                        flushQueue();
                    }
                }
            };

            NetworkRequest request = new NetworkRequest.Builder()
                    .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
                    .build();
            cm.registerNetworkCallback(request, networkCallback);
            Log.i(TAG, "Network monitor started");
        } catch (Exception e) {
            Log.e(TAG, "Network monitor error: " + e.getMessage());
        }
    }

    // ═══════════════════════════════════════════════════════════════
    // Configure
    // ═══════════════════════════════════════════════════════════════

    // ── Full config (legacy — kept for backwards compat) ──
    public void configure(String pingURL, String endURL, String userId, String vehicleId,
                          String osInfo, String routeId, String authorizationKey, String apiAuthKey) {
        configure(pingURL, endURL, userId, vehicleId, osInfo, routeId,
                authorizationKey, apiAuthKey, "");
    }

    // ── Full config with new apiAuthToken ──
    public void configure(String pingURL, String endURL, String userId, String vehicleId,
                          String osInfo, String routeId, String authorizationKey,
                          String apiAuthKey, String apiAuthToken) {
        this.pingURL = pingURL != null ? pingURL : "";
        this.endURL = endURL != null ? endURL : "";
        this.userId = userId != null ? userId : "";
        this.vehicleId = vehicleId != null ? vehicleId : "";
        if (osInfo != null && !osInfo.isEmpty()) this.osInfo = osInfo;
        this.routeId = routeId != null ? routeId : "";
        this.authorizationKey = authorizationKey != null ? authorizationKey : "";
        this.apiAuthKey = apiAuthKey != null ? apiAuthKey : "";
        this.apiAuthToken = apiAuthToken != null ? apiAuthToken : "";
        Log.i(TAG, "API configured: ping=" + this.pingURL + " user=" + this.userId);

        // Try flushing any pending requests now that config may have URLs
        if (isEnabled() && !pendingQueue.isEmpty()) {
            flushQueue();
        }
    }

    // ── Update vehicle_id at any time ──
    public void updateVehicleId(String vehicleId) {
        this.vehicleId = vehicleId != null ? vehicleId : "";
        this.routeId = vehicleId != null ? vehicleId : "";
        Log.i(TAG, "vehicle_id updated → " + this.vehicleId);
    }

    public void setRouteId(String id) { this.routeId = id != null ? id : ""; }
    public boolean isEnabled() { return !pingURL.isEmpty() && !endURL.isEmpty() && !userId.isEmpty(); }
    public boolean hasRouteId() { return routeId != null && !routeId.isEmpty(); }

    // ── Trip lifecycle — controls vehicle_id inclusion ──
    public void onTripStart() {
        includeVehicleId = true;
        Log.i(TAG, "Trip started — vehicle_id will be included in pings");
    }

    public void onTripEnd() {
        includeVehicleId = false;
        Log.i(TAG, "Trip ended — vehicle_id will NOT be included until next trip");
    }

    // ═══════════════════════════════════════════════════════════════
    // POST /ping/v2
    // ═══════════════════════════════════════════════════════════════

    public void sendPing(Location location, boolean isMoving, float speed, String activityType) {
        sendPing(location, isMoving, speed, activityType, this.routeId);
    }

    public void sendPing(Location location, boolean isMoving, float speed, String activityType, String routeId) {
        if (!isEnabled()) return;
        executor.execute(() -> {
            try {
                JSONObject locObj = new JSONObject();
                locObj.put("is_Moving", isMoving);
                locObj.put("timestamp", isoNow());
                locObj.put("latitude", location.getLatitude());
                locObj.put("longitude", location.getLongitude());
                locObj.put("speed", speed);
                locObj.put("activityType", activityType);
                

                JSONArray locArr = new JSONArray();
                locArr.put(locObj);

                JSONObject body = new JSONObject();
                body.put("user_Id", userId);
                body.put("os_Info", osInfo);
                body.put("location", locArr);

                // Only include vehicle_Id during active trip and if configured
                if (includeVehicleId && !vehicleId.isEmpty()) {
                    body.put("vehicle_Id", vehicleId);
                    body.put("route_Id", routeId != null ? routeId : this.routeId);
                }

                boolean ok = post(pingURL, body);
                if (ok) {
                    Log.d(TAG, "Ping OK: " + location.getLatitude() + "," + location.getLongitude());
                    // Success — try flushing pending queue too
                    if (!pendingQueue.isEmpty()) flushQueue();
                } else {
                    Log.d(TAG, "Ping FAIL — queued for retry");
                    enqueue(pingURL, body);
                }
            } catch (Exception e) {
                Log.e(TAG, "Ping error: " + e.getMessage());
            }
        });
    }

    // ═══════════════════════════════════════════════════════════════
    // POST /end
    // ═══════════════════════════════════════════════════════════════

    public void sendTripEnd(Location location) {
        if (!isEnabled()) return;
        if (routeId == null || routeId.isEmpty()) return;
        executor.execute(() -> {
            try {
                JSONObject body = new JSONObject();
                body.put("user_Id", userId);
                body.put("timestamp", isoNow());
                body.put("latitude", location.getLatitude());
                body.put("longitude", location.getLongitude());

                boolean ok = post(endURL, body);
                includeVehicleId = false;

                if (ok) {
                    Log.d(TAG, "Trip-end OK");
                    if (!pendingQueue.isEmpty()) flushQueue();
                } else {
                    Log.d(TAG, "Trip-end FAIL — queued for retry");
                    enqueue(endURL, body);
                }
            } catch (Exception e) {
                Log.e(TAG, "Trip-end error: " + e.getMessage());
            }
        });
    }

    // ═══════════════════════════════════════════════════════════════
    // HTTP POST
    // ═══════════════════════════════════════════════════════════════

    private boolean post(String urlStr, JSONObject body) {
        HttpURLConnection conn = null;
        try {
            URL url = new URL(urlStr);
            conn = (HttpURLConnection) url.openConnection();
            conn.setRequestMethod("POST");
            conn.setRequestProperty("Content-Type", "application/json");
            conn.setConnectTimeout(15000);
            conn.setReadTimeout(15000);
            conn.setDoOutput(true);

            if (!authorizationKey.isEmpty())
                conn.setRequestProperty("AuthorizationKey", authorizationKey);
            if (!apiAuthKey.isEmpty())
                conn.setRequestProperty("api-auth-key", apiAuthKey);
            if (!apiAuthToken.isEmpty())
                conn.setRequestProperty("api-auth-token", apiAuthToken);

            OutputStream os = conn.getOutputStream();
            os.write(body.toString().getBytes("UTF-8"));
            os.flush(); os.close();

            int code = conn.getResponseCode();
            return code >= 200 && code < 300;
        } catch (Exception e) {
            Log.e(TAG, "POST error: " + e.getMessage());
            return false;
        } finally {
            if (conn != null) conn.disconnect();
        }
    }

    private String isoNow() {
        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US);
        sdf.setTimeZone(TimeZone.getTimeZone("UTC"));
        return sdf.format(new Date());
    }
}

File: triptracking-library/ios/Sources/triptracking/Services/TripTrackerAPIService.swift
import Foundation
import CoreLocation
import UIKit
import Network

// MARK: - API Configuration

public struct TripTrackerAPIConfig {
    public var pingURL: String = ""
    public var endURL: String = ""
    public var userId: String = ""
    public var vehicleId: String = ""       // Optional — can be empty
    public var osInfo: String = ""
    public var routeId: String = ""
    public var authorizationKey: String = ""
    public var apiAuthKey: String = ""        // Legacy — kept for backwards compat
    public var apiAuthToken: String = ""      // New header: api-auth-token

    public var isConfigured: Bool { !pingURL.isEmpty && !endURL.isEmpty && !userId.isEmpty }

    public init() { self.osInfo = "iOS \(UIDevice.current.systemVersion)" }

    public init(from dict: [String: Any]) {
        self.init()
        if let v = dict["pingURL"] as? String         { pingURL = v }
        if let v = dict["endURL"] as? String           { endURL = v }
        if let v = dict["userId"] as? String           { userId = v }
        if let v = dict["vehicleId"] as? String        { vehicleId = v }
        if let v = dict["osInfo"] as? String           { osInfo = v }
        if let v = dict["routeId"] as? String          { routeId = v }
        if let v = dict["authorizationKey"] as? String { authorizationKey = v }
        if let v = dict["apiAuthKey"] as? String       { apiAuthKey = v }
        if let v = dict["apiAuthToken"] as? String     { apiAuthToken = v }
    }
}

// MARK: - API Service

public final class TripTrackerAPIService {
    public static let shared = TripTrackerAPIService()
    private init() {
        loadPendingQueue()
        startNetworkMonitor()
    }

    public var config = TripTrackerAPIConfig()
    public var isEnabled: Bool { config.isConfigured }

    /// Whether to include vehicle_Id in outgoing payloads.
    /// Automatically set to true on trip start, false on trip end.
    public var includeVehicleId: Bool = false

    private let session: URLSession = {
        let cfg = URLSessionConfiguration.default
        cfg.timeoutIntervalForRequest = 15
        return URLSession(configuration: cfg)
    }()

    // ═══════════════════════════════════════════════════════════════
    // Retry Queue — persists failed requests to disk, resends on network
    // ═══════════════════════════════════════════════════════════════

    private var pendingQueue: [[String: Any]] = []  // [{url, body}]
    private let queueLock = NSLock()
    private let maxQueueSize = 500
    private var isFlushing = false

    private var queueFileURL: URL? {
        FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)
            .first?.appendingPathComponent("triptracker_pending_api.json")
    }

    private func loadPendingQueue() {
        guard let url = queueFileURL,
              let data = try? Data(contentsOf: url),
              let arr = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]] else { return }
        pendingQueue = arr
        print("📡  TripTracker API queue loaded: \(arr.count) pending requests")
    }

    private func savePendingQueue() {
        guard let url = queueFileURL else { return }
        queueLock.lock()
        defer { queueLock.unlock() }
        if let data = try? JSONSerialization.data(withJSONObject: pendingQueue) {
            try? data.write(to: url)
        }
    }

    private func enqueue(url: String, body: [String: Any]) {
        queueLock.lock()
        pendingQueue.append(["url": url, "body": body, "ts": Date().timeIntervalSince1970])
        // Trim oldest if over limit
        if pendingQueue.count > maxQueueSize {
            pendingQueue.removeFirst(pendingQueue.count - maxQueueSize)
        }
        queueLock.unlock()
        savePendingQueue()
        print("📡  TripTracker API queued (total: \(pendingQueue.count)) — will retry when online")
    }

    /// Flush all pending requests. Called when network becomes available.
    public func flushQueue() {
        guard !isFlushing else { return }
        queueLock.lock()
        let items = pendingQueue
        queueLock.unlock()
        guard !items.isEmpty else { return }

        isFlushing = true
        print("📡  TripTracker API flushing \(items.count) pending requests…")

        DispatchQueue.global(qos: .utility).async { [weak self] in
            var successCount = 0
            for item in items {
                guard let urlStr = item["url"] as? String,
                      let body = item["body"] as? [String: Any] else { continue }

                let ok = self?.postSync(url: urlStr, body: body) ?? false
                if ok {
                    successCount += 1
                    // Remove from queue
                    self?.queueLock.lock()
                    if let idx = self?.pendingQueue.firstIndex(where: { ($0["ts"] as? Double) == (item["ts"] as? Double) }) {
                        self?.pendingQueue.remove(at: idx)
                    }
                    self?.queueLock.unlock()
                } else {
                    // Still offline — stop flushing
                    break
                }
            }
            self?.savePendingQueue()
            self?.isFlushing = false
            let remaining = self?.pendingQueue.count ?? 0
            print("📡  TripTracker API flush done: \(successCount) sent, \(remaining) remaining")
        }
    }

    /// Number of pending requests in queue
    public var pendingCount: Int { pendingQueue.count }

    // ═══════════════════════════════════════════════════════════════
    // Network Monitor — auto-flush when connectivity returns
    // ═══════════════════════════════════════════════════════════════

    private var networkMonitor: Any?  // NWPathMonitor (stored as Any to avoid import issues)

    private func startNetworkMonitor() {
        if #available(iOS 12.0, *) {
            let monitor = NWPathMonitor()
            monitor.pathUpdateHandler = { [weak self] path in
                if path.status == .satisfied && (self?.pendingQueue.isEmpty == false) {
                    print("📡  TripTracker API network restored — flushing pending queue")
                    self?.flushQueue()
                }
            }
            monitor.start(queue: DispatchQueue.global(qos: .utility))
            networkMonitor = monitor
        }
    }

    // ═══════════════════════════════════════════════════════════════
    // Public API
    // ═══════════════════════════════════════════════════════════════

    public func updateVehicleId(_ vehicleId: String) {
        config.vehicleId = vehicleId
        config.routeId = vehicleId
        print("📡  TripTracker API vehicle_id updated → \(vehicleId)")
    }

    // ── Called on trip start to start including vehicle_id ──
    public func onTripStart() {
        includeVehicleId = true
    }

    // ── Called on trip end to stop including vehicle_id ──
    public func onTripEnd() {
        includeVehicleId = false
    }

    // POST /ping/v2
    public func sendPing(location: CLLocation, isMoving: Bool, speed: Float, activityType: String, routeId: String? = nil) {
        if(isEnabled) {
            print("📡 TripTracker Sending ping for location: \(location.coordinate.latitude),\(location.coordinate.longitude)")
        }else{
            print("📡 TripTracker Ping NOT sent because API config is incomplete")
        }
        var body: [String: Any] = [
            "user_Id": config.userId,
            "os_Info": config.osInfo,
            "location": [[
                "is_Moving": isMoving,
                "timestamp": ISO8601DateFormatter().string(from: Date()),
                "latitude": location.coordinate.latitude,
                "longitude": location.coordinate.longitude,
                "speed": speed,
                "activityType": activityType,
                "route_Id": includeVehicleId ? routeId ?? config.vehicleId : ""
                
            ]]
        ]
        print("📡  TripTracker API ping route: \(body) (vehicleId included: \(includeVehicleId))")
        
        // Only include vehicle_Id during active trip and if configured
        if includeVehicleId && !config.vehicleId.isEmpty {
            body["vehicle_Id"] = config.vehicleId
        }
        post(url: config.pingURL, body: body) { ok in
            print("📡 TripTrackerAPI ping \(ok ? "OK" : "FAIL"): \(location.coordinate.latitude),\(location.coordinate.longitude)")
        }
    }

    // POST /ping/v2 batch
    public func sendPingBatch(locations: [(CLLocation, Bool, Float, String, Date)], routeId: String? = nil) {
        if(isEnabled) {
            print("📡 TripTracker Sending batch ping for \(locations.count) locations")
        } else {
            print("📡 TripTracker Batch ping NOT sent because API config is incomplete")
        }
        guard isEnabled, !locations.isEmpty else { return }
        let fmt = ISO8601DateFormatter()
        let arr: [[String: Any]] = locations.map { loc, moving, spd, activity, ts in
            ["is_Moving": moving, "timestamp": fmt.string(from: ts),
             "latitude": loc.coordinate.latitude, "longitude": loc.coordinate.longitude,
             "speed": spd, "activityType": activity, "route_Id": includeVehicleId ? routeId ?? config.vehicleId : ""]
        }
        print("📡  TripTracker API ping route: \(includeVehicleId ? routeId ?? config.vehicleId : "") (vehicleId included: \(includeVehicleId))")
        var body: [String: Any] = ["user_Id": config.userId, "os_Info": config.osInfo, "location": arr]
        if includeVehicleId && !config.vehicleId.isEmpty {
            body["vehicle_Id"] = config.vehicleId
        }
        post(url: config.pingURL, body: body) { ok in
            print("📡  TripTracker API batch (\(locations.count)): \(ok ? "OK" : "FAIL")")
        }
    }

    // POST /end — vehicle_id NOT included after trip end
    public func sendTripEnd(location: CLLocation) {
        if(isEnabled) {
            print("📡 TripTracker Sending trip end for location: \(location.coordinate.latitude),\(location.coordinate.longitude)")
        }else{
            print("📡 TripTracker Trip end NOT sent because API config is incomplete")
        }
        let body: [String: Any] = [
            "user_Id": config.userId,
            "timestamp": ISO8601DateFormatter().string(from: Date()),
            "latitude": location.coordinate.latitude,
            "longitude": location.coordinate.longitude
        ]
        post(url: config.endURL, body: body) { [weak self] ok in
            print("📡 TripTracker API trip-end \(ok ? "OK" : "FAIL")")
            // Stop including vehicle_id after trip end
            self?.includeVehicleId = false
            if !ok {
                DispatchQueue.global().asyncAfter(deadline: .now() + 5) {
                    self?.post(url: self?.config.endURL ?? "", body: body, completion: nil)
                }
            }
        }
    }

    public func setRouteId(_ id: String) { config.routeId = id }

    // ═══════════════════════════════════════════════════════════════
    // HTTP — post with automatic retry queue on failure
    // ═══════════════════════════════════════════════════════════════

    private func postWithRetry(url: String, body: [String: Any], completion: ((Bool) -> Void)?) {
        post(url: url, body: body) { [weak self] ok in
            if !ok {
                self?.enqueue(url: url, body: body)
            } else if self?.pendingQueue.isEmpty == false {
                // Success — try flushing pending queue too
                self?.flushQueue()
            }
            completion?(ok)
        }
    }

    private func post(url: String, body: [String: Any], completion: ((Bool) -> Void)?) {
        guard let url = URL(string: url) else { completion?(false); return }
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if !config.authorizationKey.isEmpty { req.setValue(config.authorizationKey, forHTTPHeaderField: "AuthorizationKey") }
        if !config.apiAuthKey.isEmpty { req.setValue(config.apiAuthKey, forHTTPHeaderField: "api-auth-key") }
        if !config.apiAuthToken.isEmpty { req.setValue(config.apiAuthToken, forHTTPHeaderField: "api-auth-token") }
        do { req.httpBody = try JSONSerialization.data(withJSONObject: body) } catch { completion?(false); return }
        session.dataTask(with: req) { data, response, error in
            let code = (response as? HTTPURLResponse)?.statusCode ?? 0
            completion?((200...299).contains(code))
        }.resume()
    }

    /// Synchronous POST — used by flush queue on background thread
    private func postSync(url urlStr: String, body: [String: Any]) -> Bool {
        guard let url = URL(string: urlStr) else { return false }
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if !config.authorizationKey.isEmpty { req.setValue(config.authorizationKey, forHTTPHeaderField: "AuthorizationKey") }
        if !config.apiAuthKey.isEmpty { req.setValue(config.apiAuthKey, forHTTPHeaderField: "api-auth-key") }
        if !config.apiAuthToken.isEmpty { req.setValue(config.apiAuthToken, forHTTPHeaderField: "api-auth-token") }
        guard let httpBody = try? JSONSerialization.data(withJSONObject: body) else { return false }
        req.httpBody = httpBody

        let sem = DispatchSemaphore(value: 0)
        var success = false
        session.dataTask(with: req) { _, response, _ in
            let code = (response as? HTTPURLResponse)?.statusCode ?? 0
            success = (200...299).contains(code)
            sem.signal()
        }.resume()
        sem.wait()
        return success
    }
}
